pkg/cloudops/aws/aws.go

```diff
@@ -0,0 +1,257 @@
+package aws
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/libopenstorage/openstorage/pkg/cloudops"
+	"github.com/libopenstorage/openstorage/pkg/util"
+	"github.com/sirupsen/logrus"
+)
+
+// ErrAWSEnvNotAvailable is the error type when aws credentials are not set
+var ErrAWSEnvNotAvailable = fmt.Errorf("AWS credentials are not set in environment")
+
+type awsOps struct {
+	instInfo    *cloudops.InstanceInfo
+	ec2         *ec2.EC2
+	autoscaling *autoscaling.AutoScaling
+}
+
+// NewAWSps returns an instance of the AWS cloud ops implementation
+func NewAWSOps() (cloudops.Ops, error) {
+	ops := &awsOps{
+		instInfo: &cloudops.InstanceInfo{},
+	}
+	err := ops.populateBasicInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	ops.instInfo.Region = ops.instInfo.Zone[:len(ops.instInfo.Zone)-1]
+
+	ops.ec2 = ec2.New(
+		session.New(
+			&aws.Config{
+				Region:      &ops.instInfo.Region,
+				Credentials: credentials.NewEnvCredentials(),
+			},
+		),
+	)
+
+	ops.autoscaling = autoscaling.New(
+		session.New(
+			&aws.Config{
+				Region:      &ops.instInfo.Region,
+				Credentials: credentials.NewEnvCredentials(),
+			},
+		),
+	)
+
+	return ops, nil
+}
+
+func (a *awsOps) InspectSelf() (*cloudops.InstanceInfo, error) {
+	inst, err := DescribeInstanceByID(a.ec2, a.instInfo.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	a.instInfo.Labels = labelsFromTags(inst.Tags)
+	return a.instInfo, nil
+}
+
+func (a *awsOps) InspectSelfInstanceGroup() (*cloudops.InstanceGroupInfo, error) {
+	selfInfo, err := a.InspectSelf()
+	if err != nil {
+		return nil, err
+	}
+
+	for tag, value := range selfInfo.Labels {
+		// https://docs.aws.amazon.com/autoscaling/ec2/userguide/autoscaling-tagging.html#tag-lifecycle
+		if tag == "aws:autoscaling:groupName" {
+			input := &autoscaling.DescribeAutoScalingGroupsInput{
+				AutoScalingGroupNames: []*string{
+					aws.String(value),
+				},
+			}
+
+			result, err := a.autoscaling.DescribeAutoScalingGroups(input)
+			if err != nil {
+				if aerr, ok := err.(awserr.Error); ok {
+					return nil, aerr
+				} else {
+					return nil, err
+				}
+			}
+
+			if len(result.AutoScalingGroups) != 1 {
+				return nil, fmt.Errorf("DescribeAutoScalingGroups (%v) returned %v groups, expect 1",
+					value, len(result.AutoScalingGroups))
+			}
+
+			group := result.AutoScalingGroups[0]
+			if group.MinSize == nil || group.MaxSize == nil {
+				return nil, fmt.Errorf("Autoscaling group: %s does not have min or max set", value)
```

Comment:
we should have MinSize and MaxSize as pointers to int and the callers can decide what to do if they are nil.
Lets not return an error here.

Reply:
done